cassiopeia/preprocess/filter_utils.py

```diff
@@ -0,0 +1,232 @@
+        if x.shape[0] > 0:
+            umi_per_ibc = np.concatenate([umi_per_ibc, np.array(x)])
+
+    # Count UMI per cellBC
+    umi_per_cbc = (
+        moleculetable.groupby(["cellBC"])
+        .agg({"UMI": "count"})
+        .sort_values("UMI", ascending=False)["UMI"]
+    )
+
+    return (
+        np.array(moleculetable["ReadCount"]),
+        umi_per_ibc,
+        np.array(umi_per_cbc),
+    )
+
+
+def filter_cellbcs(
+    moleculetable: pd.DataFrame, umiCountThresh: int = 10, verbose: bool = False
+) -> pd.DataFrame:
+    """
+    Filter out cell barcodes that have too few UMIs.
+
+    Filters out cell barcodes that have a number of unique UMIs <= umiCountThresh.
+    Assumes that UMIs have been resolved, i.e. that each UMI only appears once.
+
+    Args:
+        moleculetable: A DataFrame of alignments to be filtered
+        umiCountThresh: The minimum number of UMIs needed for a cellBC to be
+            included in the filtered DataFrame
+        verbose: Indicates whether to log the number of cellBCs and UMIs
+            remaining after filtering
+
+    Returns:
+        n_moleculetable: A filtered DataFrame of alignments.
+    """
+
+    # Create a cell-filter dictionary for hash lookup lmter on when filling
+    # in the table
+    cell_filter = {}
+
+    for n, group in tqdm(
+        moleculetable.groupby(["cellBC"]), desc="Filtering cellBCs for UMIs"
+    ):
+        if group.shape[0] <= umiCountThresh:
+            cell_filter[n] = "bad"
+            # tooFewUMI_UMI.append(group.shape[0])
+        else:
+            cell_filter[n] = "good"
+
+    # apply the filter using the hash table created above
+    moleculetable["status"] = moleculetable["cellBC"].map(cell_filter)
+
+    # filter based on status & reindex
+    n_moleculetable = moleculetable[(moleculetable["status"] == "good")]
+    n_moleculetable.index = [i for i in range(n_moleculetable.shape[0])]
+
+    if verbose:
+        lg_utils.generate_log_output(n_moleculetable)
+
+    return n_moleculetable
+
+
+def filter_umis(
+    moleculetable: pd.DataFrame,
+    readCountThresh: int = 100,
+    verbose: bool = False,
+) -> pd.DataFrame:
+    """
+    Filters out UMIs with too few reads.
+
+    Filters out all UMIs with a read count <= readCountThresh.
+
+    Args:
+        moleculetable: A DataFrame of alignments to be filtered
+        readCountThresh: The minimum read count needed for a UMI to be
+            included in the filtered DataFrame
+        verbose: Indicates whether to log the number of cellBCs and UMIs
+            remaining after filtering
+
+    Returns:
+        n_moleculetable: A filtered DataFrame of alignments.
+    """
+
+    # filter based on status & reindex
+    n_moleculetable = moleculetable[
+        moleculetable["ReadCount"] > readCountThresh
+    ]
+    n_moleculetable.index = [i for i in range(n_moleculetable.shape[0])]
+
+    if verbose:
+        lg_utils.generate_log_output(n_moleculetable)
+
+    return n_moleculetable
+
+
```

Comment:
See my comments above for the version in `ProcessingPipeline.process.filter_utils.py`